### ISSUE TITLE
State seems shared by component instances

### ISSUE THREAD

**Issue by italonascimento** [recyclejs/recycle#20]
I was facing a weird behavior with sibling instances of a same component, and now I think I was able to isolate the issue in this simple exemple:

```js
const Exemple = recycle({
  initialState: ({
    myBoolean: false,
  }),

  update(sources) {
    return [
      sources.lifecycle
        .filter(e => e === 'componentDidMount')
        .reducer(state => {
          state.myBoolean = !state.myBoolean
          return state
        }),
    ]
  },

  view(props, state) {
    return (
      <p>{state.myBoolean.toString()}</p>
    )
  }
})
```

I would expect all instances of this component to negate `myBoolean` on mount, making it `true`, as the initial state defines it `false`. But instead, each instance gets a different value, as they were negating the same state reference.

If I do:
```jsx
<Exemple/>
<Exemple/>
<Exemple/>
```

I get:
```jsx
<p>false</p>
<p>true</p>
<p>false</p>
```

Then, if I navigate and come back to the page, it inverts:
```jsx
<p>true</p>
<p>false</p>
<p>true</p>
```

**Comment by italonascimento** [comment]
Found the issue to be the mutation of the `state` inside `reducer`. As the object reference been mutated is the `initialState`, it affects the other instances to be mounted.

This fixes the problem:

```js
.reducer(state => ({
  ...state,
  myBoolean = !state.myBoolean
}))
```

It's weird though, as the docs spend some effort convincing us it's ok to mutate inside `reducer`.

**Comment by domagojk** [comment]
Thank you for isolating this!
You're correct, there was a bug - all component instances shared the same `initialState`.
This should be fixed in `v 2.2.3` and you should see expected values without modifying `reducer`.

Note that all of this works only for "flat state". Deeply nested objects are not supported.

**Comment by italonascimento** [comment]
Nice!
Thanks for the quick response. :)
